blocks/model.py

```diff
@@ -1,75 +1,105 @@
-
 from blocks.graph import ComputationGraph
 from blocks.select import Selector
 from blocks.filter import get_brick
 
 logger = logging.getLogger(__name__)
 
-multiple_message = """
-
-Model with multiple outputs are currently only partially supported \
-in Blocks. For instance a call of 'get_objective' will crash. \
-Contact Blocks developers for more details.
-"""
-
-
-@add_metaclass(ABCMeta)
-class AbstractModel(object):
-    """A parameterized entity trained in the main loop.
 
-    A model is a parameterized entity the user trains a in a main loop.
-    The following are traits of every model:
-
-    * It has parameters and supports a way to access them. In addition
-      to returning handles to parameter objects it can return their values
-      as numpy arrays and set their values to given numpy arrays.
-
-    * It has an optimality objective.
-
-    * It can be serialized and deserialized by mean of pickling.
-
-    * It might have bricks as its components.
-
-    This class provides an interface for models. For experiments use
-    a subclass, e.g. the :class:`Model`.
+class Model(ComputationGraph):
+    """Handles annotations in Blocks-built computation graphs.
+
+    Use this class to handle your Blocks-created computation graph.
+
+    Examples
+    --------
+    >>> from theano import tensor
+    >>> from blocks.bricks import MLP, Tanh
+    >>> x = tensor.matrix('x')
+    >>> mlp = MLP([Tanh(), Tanh()], [10, 10, 10])
+    >>> y = mlp.apply(x)
+    >>> model = Model(y)
+    >>> # With model you can get access to the brick hierarchy
+    >>> model.get_top_bricks() #doctest: +ELLIPSIS
+    [<blocks.bricks.MLP object at ...]
+    >>> # With model you get "hierarchical" names for the parameters,
+    >>> # which encode the position of the owning brick in the
+    >>> # brick hierarchy.
+    >>> model.get_parameter_dict() #doctest: +NORMALIZE_WHITESPACE
+    OrderedDict([('/mlp/linear_1.b', b), ('/mlp/linear_0.b', b),
+    ('/mlp/linear_0.W', W), ('/mlp/linear_1.W', W)])
 
     """
-    @abstractmethod
+    def __init__(self, *args, **kwargs):
+        bricks = [get_brick(var) for var
+                  in self.variables + self.scan_variables if get_brick(var)]
+        children = set(chain(*(brick.children for brick in bricks)))
+        # Quadratic complexity: we should not have thousands of
+        # top-level bricks.
+        self.top_bricks = []
+        for brick in bricks:
+            if brick not in children and brick not in self.top_bricks:
+                self.top_bricks.append(brick)
+        names = Counter([brick.name for brick in self.top_bricks])
+        repeated_names = [name for name, count in names.items() if count > 1]
+        if repeated_names:
+            raise ValueError("top bricks with the same name:"
+                             " {}".format(', '.join(repeated_names)))
+        brick_parameter_names = {
+            v: k for k, v in Selector(
+                self.top_bricks).get_parameters().items()}
```

Comment:
Oh sure, the elements are actually not copied... I reverted it back to items() :)